Compilation failure with latest glib

When building on latest fedora rawhide hitting this error:
```
[27/337] Compiling C object subprojects/wocky/wocky/libwocky-telepathy-0.19.0.so.0.19.0.p/wocky-auth-handler.c.o
FAILED: subprojects/wocky/wocky/libwocky-telepathy-0.19.0.so.0.19.0.p/wocky-auth-handler.c.o 
cc -Isubprojects/wocky/wocky/libwocky-telepathy-0.19.0.so.0.19.0.p -Isubprojects/wocky/wocky -I../subprojects/wocky/wocky -Isubprojects/wocky -I../subprojects/wocky -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/libxml2 -I/usr/include/libsoup-2.4 -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -std=c99 -g -D_GNU_SOURCE -DHAVE_CONFIG_H -DG_LOG_USE_STRUCTURED '-DLOCALE_DIR="/usr/local/share/locale"' -Wdeclaration-after-statement -Wimplicit-function-declaration -Wshadow -Wstrict-prototypes -Wmissing-prototypes -Wsign-compare -Wnested-externs -Wpointer-arith -Wformat -Wformat-security -Winit-self -Werror -Wno-missing-field-initializers -Wno-error=missing-field-initializers -Wno-cast-function-type -Wno-error=cast-function-type -Wno-unused-parameter -Wno-error=unused-parameter -fPIC -pthread '-DG_LOG_DOMAIN="wocky"' -DWOCKY_COMPILATION -MD -MQ subprojects/wocky/wocky/libwocky-telepathy-0.19.0.so.0.19.0.p/wocky-auth-handler.c.o -MF subprojects/wocky/wocky/libwocky-telepathy-0.19.0.so.0.19.0.p/wocky-auth-handler.c.o.d -o subprojects/wocky/wocky/libwocky-telepathy-0.19.0.so.0.19.0.p/wocky-auth-handler.c.o -c ../subprojects/wocky/wocky/wocky-auth-handler.c
In file included from /usr/include/glib-2.0/glib/gthread.h:32,
                 from /usr/include/glib-2.0/glib/gasyncqueue.h:32,
                 from /usr/include/glib-2.0/glib.h:32,
                 from ../subprojects/wocky/wocky/wocky-auth-handler.h:8,
                 from ../subprojects/wocky/wocky/wocky-auth-handler.c:5:
../subprojects/wocky/wocky/wocky-auth-handler.c: In function ‘wocky_auth_handler_get_type’:
/usr/include/glib-2.0/glib/gatomic.h:112:5: error: argument 2 of ‘__atomic_load’ discards ‘volatile’ qualifier [-Werror=incompatible-pointer-types]
  112 |     __atomic_load (gapg_temp_atomic, &gapg_temp_newval, __ATOMIC_SEQ_CST); \
      |     ^~~~~~~~~~~~~
/usr/include/glib-2.0/glib/gthread.h:260:7: note: in expansion of macro ‘g_atomic_pointer_get’
  260 |     (!g_atomic_pointer_get (location) &&                             \
      |       ^~~~~~~~~~~~~~~~~~~~
../subprojects/wocky/wocky/wocky-auth-handler.c:13:7: note: in expansion of macro ‘g_once_init_enter’
   13 |   if (g_once_init_enter (&g_define_type_id__volatile))
      |       ^~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors

```
Need to migrate to `G_DEFINE_INTERFACE`

--- a/tests/wocky-test-sasl-auth-server.c
+++ b/tests/wocky-test-sasl-auth-server.c
@@ -784,9 +784,30 @@ handle_auth (TestSaslAuthServer *self, WockyStanza *stanza)
           GError *err = NULL;
           gchar *cb64 = NULL;
           GByteArray *cb = g_byte_array_new ();
+          const gchar *cb_str = g_getenv ("WOCKY_CHANNEL_BINDING_TYPE");
+          int g_tls_cb_t = G_TLS_CHANNEL_BINDING_TLS_UNIQUE;
+          WockyTLSBindingType cb_type = WOCKY_TLS_BINDING_TLS_UNIQUE;
 
-          if (!g_tls_connection_get_channel_binding_data ((GTlsConnection *) ios,
-                  G_TLS_CHANNEL_BINDING_TLS_UNIQUE, cb, &err))
+          /* by default it is tls-unique, but if it's something else what we
+           * has some knowledge about - let's use it, otherwise we fail */
+          if (g_str_has_prefix ((gchar *) response, "p="))
+            {
+              GEnumClass *gec = g_type_class_ref (WOCKY_TYPE_TLS_BINDING_TYPE);
+              if (g_str_has_prefix ((gchar *) (response + 2), "tls-server-end-point,"))
+                {
+                  cb_type = WOCKY_TLS_BINDING_TLS_SERVER_END_POINT;
+                  g_tls_cb_t = G_TLS_CHANNEL_BINDING_TLS_SERVER_END_POINT;
+                }
+              else if (g_str_has_prefix ((gchar *) (response + 2), "tls-exporter,"))
+                {
+                  cb_type = WOCKY_TLS_BINDING_TLS_EXPORTER;
+                  g_tls_cb_t = 100500;
+                }
+              cb_str = g_enum_get_value (gec, cb_type)->value_nick;
+              g_type_class_unref (gec);
+            }
+
+          if (!g_tls_connection_get_channel_binding_data ((GTlsConnection *) ios, g_tls_cb_t, cb, &err))
             {
               g_assert_error (err, G_TLS_CHANNEL_BINDING_ERROR,
                   G_TLS_CHANNEL_BINDING_ERROR_NOT_IMPLEMENTED);
@@ -803,9 +824,10 @@ handle_auth (TestSaslAuthServer *self, WockyStanza *stanza)
               else if (priv->problem == SERVER_PROBLEM_MANGLED_BINDING_FLAG
                     || priv->problem == SERVER_PROBLEM_SCRAMBLED_BINDING)
                 {
-                  gchar *r;
-                  g_assert_true (g_str_has_prefix ((gchar *) response, "p=tls-unique,"));
-                  r = g_strdup_printf ("n%s", response + 12);
+                  gchar *r = g_strdup_printf ("p=%s,", cb_str);
+                  g_assert_true (g_str_has_prefix ((gchar *) response, r));
+                  g_free (r);
+                  r = g_strdup_printf ("n%s", response + 2 + strlen (cb_str));
                   g_free (response);
                   response = (guchar *) r;
                 }
@@ -813,7 +835,7 @@ handle_auth (TestSaslAuthServer *self, WockyStanza *stanza)
               cb64 = g_base64_encode (cb->data, cb->len);
               g_byte_array_unref (cb);
               g_object_set (G_OBJECT (priv->scram),
-                        "cb-type", WOCKY_TLS_BINDING_TLS_UNIQUE,
+                        "cb-type", cb_type,
                         "cb-data", cb64,
                         NULL);
               g_debug ("TLS binding: %s", cb64);

--- a/wocky/wocky-auth-handler.c
+++ b/wocky/wocky-auth-handler.c
@@ -5,34 +5,13 @@
 #include "wocky-auth-handler.h"
 #include "wocky-auth-registry.h"
 
-GType
-wocky_auth_handler_get_type (void)
+typedef WockyAuthHandlerIface WockyAuthHandlerInterface;
+
+G_DEFINE_INTERFACE (WockyAuthHandler, wocky_auth_handler, G_TYPE_OBJECT)
+
+static void
+wocky_auth_handler_default_init (WockyAuthHandlerInterface *iface)
 {
-  static volatile gsize g_define_type_id__volatile = 0;
-
-  if (g_once_init_enter (&g_define_type_id__volatile))
-    {
-      const GTypeInfo info =
-      {
-        /* class_size */ sizeof (WockyAuthHandlerIface),
-        /* base_init */ NULL,
-        /* base_finalize */ NULL,
-        /* class_init */ NULL,
-        /* class_finalize */ NULL,
-        /* class_data */ NULL,
-        /* instance_size */ 0,
-        /* n_preallocs */ 0,
-        /* instance_init */ NULL,
-        /* value_table */ NULL
-      };
-      GType g_define_type_id = g_type_register_static (
-          G_TYPE_INTERFACE, "WockyAuthHandler", &info, 0);
-
-      g_type_interface_add_prerequisite (g_define_type_id, G_TYPE_OBJECT);
-      g_once_init_leave (&g_define_type_id__volatile, g_define_type_id);
-    }
-
-  return g_define_type_id__volatile;
 }
 
 /**

--- a/wocky/wocky-connector.c
+++ b/wocky/wocky-connector.c
@@ -373,6 +373,7 @@ abort_connect_error (WockyConnector *connector,
     }
 
   g_task_return_error (priv->task, g_error_copy (*error));
+  complete_operation (connector);
 }
 
 static void
@@ -395,6 +396,7 @@ abort_connect (WockyConnector *connector,
     }
 
   g_task_return_error (priv->task, g_error_copy (error));
+  complete_operation (connector);
 }
 
 static void

--- a/wocky/wocky-muc.c
+++ b/wocky/wocky-muc.c
@@ -1412,12 +1412,16 @@ get_message_sender (WockyMuc *muc,
 static GDateTime *
 extract_timestamp (WockyNode *msg)
 {
-  WockyNode *x = wocky_node_get_child_ns (msg, "x", WOCKY_XMPP_NS_DELAY);
+  WockyNode *x = wocky_node_get_child_ns (msg, "x", WOCKY_NS_JABBER_DELAY);
   GDateTime *stamp = NULL;
+
+  if (x == NULL)
+    x = wocky_node_get_child_ns (msg, "delay", WOCKY_XMPP_NS_DELAY);
 
   if (x != NULL)
     {
       const gchar *tm = wocky_node_get_attribute (x, "stamp");
+      GTimeZone *tz = g_time_zone_new_utc ();
 
       /* These timestamps do not contain a timezone, but are understood to be
        * in GMT. They're in the format yyyymmddThhmmss, so if we append 'Z'
@@ -1425,9 +1429,10 @@ extract_timestamp (WockyNode *msg)
        */
       if (tm != NULL)
         {
-          if ((stamp = g_date_time_new_from_iso8601 (tm, NULL)) == NULL)
+          if ((stamp = g_date_time_new_from_iso8601 (tm, tz)) == NULL)
             DEBUG ("Malformed date string '%s' for " WOCKY_XMPP_NS_DELAY, tm);
         }
+      g_time_zone_unref (tz);
     }
 
   return stamp;

--- a/wocky/wocky-namespaces.h
+++ b/wocky/wocky-namespaces.h
@@ -65,8 +65,11 @@
 #define WOCKY_XMPP_NS_EVENT \
   "jabber:x:event"
 
-#define WOCKY_XMPP_NS_DELAY \
+#define WOCKY_NS_JABBER_DELAY \
   "jabber:x:delay"
+
+#define WOCKY_XMPP_NS_DELAY \
+  "urn:xmpp:delay"
 
 #define WOCKY_XMPP_NS_STANZAS \
   "urn:ietf:params:xml:ns:xmpp-stanzas"

--- a/wocky/wocky-sasl-auth.c
+++ b/wocky/wocky-sasl-auth.c
@@ -289,9 +289,9 @@ auth_failed (WockySaslAuth *sasl, gint code, const gchar *format, ...)
 
   error = g_error_new_literal (WOCKY_AUTH_ERROR, code, message);
 
-  g_task_return_error (t, g_error_copy (error));
-
   wocky_auth_registry_failure (priv->auth_registry, error);
+
+  g_task_return_error (t, g_error_copy (error));
 
   g_object_unref (t);
 

--- a/wocky/wocky-tls.c
+++ b/wocky/wocky-tls.c
@@ -286,6 +286,33 @@ wocky_tls_session_verify_peer (WockyTLSSession    *session,
    * and then translate GIO validation error to Wocky language with the above dictionary  */
   peer_cert_status = g_tls_connection_get_peer_certificate_errors (G_TLS_CONNECTION (session));
 
+  /* Force-override identity check if required */
+  if (peer_cert_status & check & G_TLS_CERTIFICATE_BAD_IDENTITY
+      && extra_identities && *extra_identities)
+    {
+      GTlsCertificate *cert = g_tls_connection_get_peer_certificate (
+          G_TLS_CONNECTION (session));
+
+      DEBUG ("Checking extra identities");
+      for (int i=0; extra_identities[i] != NULL; i++)
+        {
+          GSocketConnectable *xid = g_network_address_new (extra_identities[i], 0);
+          GTlsCertificateFlags flags = g_tls_certificate_verify (cert,
+              xid, NULL);
+
+          g_object_unref (xid);
+
+          if (!(flags & G_TLS_CERTIFICATE_BAD_IDENTITY))
+            {
+              /* The certificate matches extra identity therefore disable
+               * identity checks */
+              DEBUG ("Certificate identity matches extra id %s", extra_identities[i]);
+              check &= ~G_TLS_CERTIFICATE_BAD_IDENTITY;
+              break;
+            }
+        }
+    }
+
   if (peer_cert_status & check)
     { /* gio cert checking can return multiple errors bitwise &ed together    *
        * but we are realy only interested in the "most important" error:      */
@@ -293,7 +320,7 @@ wocky_tls_session_verify_peer (WockyTLSSession    *session,
       for (x = 0; status_map[x].gio != 0; x++)
         {
           DEBUG ("checking gio error %d", status_map[x].gio);
-          if (peer_cert_status & status_map[x].gio)
+          if (peer_cert_status & check & status_map[x].gio)
             {
               DEBUG ("gio error %d set", status_map[x].gio);
               *status = status_map[x].wocky;

